packages/vm/lib/evm/opFns.ts

```diff
@@ -1237,6 +1261,98 @@ function subMemUsage(runState: RunState, offset: BN, length: BN) {
+
+  const addressStr = addressToBuffer(address).toString('hex')
+
+  // Cold
+  if (!runState.accessedAddresses.has(addressStr)) {
+    runState.accessedAddresses.add(addressStr)
+
+    // CREATE, CREATE2 opcodes have the address warmed for free.
+    if (baseFee) {
+      runState.eei.useGas(
+        new BN(runState._common.param('gasPrices', 'coldaccountaccess') - baseFee),
+      )
+    }
+    // Warm
+  } else if (baseFee) {
+    runState.eei.useGas(new BN(runState._common.param('gasPrices', 'warmstorageread') - baseFee))
+  }
+}
+
+/**
+ * Adds (address, key) to accessedStorage tuple set if not already included.
+ * Adjusts cost incurred for executing opcode based on whether storage read
+ * is warm/cold. (EIP 2929)
+ * @param {RunState} runState
+ * @param {Buffer} key (to storage slot)
+ */
+function accessStorageEIP2929(runState: RunState, key: Buffer, isSstore: boolean) {
+  if (!runState._common.eips().includes(2929)) return
+
+  const keyStr = key.toString('hex')
+  const baseFee = !isSstore ? runState._common.param('gasPrices', 'sload') : 0
+  const address = runState.eei.getAddress().toString('hex')
+  const keysAtAddress = runState.accessedStorage.get(address)
+
+  // Cold (SLOAD and SSTORE)
+  if (!keysAtAddress) {
+    runState.accessedStorage.set(address, new Set())
+    // @ts-ignore Set Object is possibly 'undefined'
+    runState.accessedStorage.get(address).add(keyStr)
+    runState.eei.useGas(new BN(runState._common.param('gasPrices', 'coldsload') - baseFee))
+  } else if (keysAtAddress && !keysAtAddress.has(keyStr)) {
+    keysAtAddress.add(keyStr)
+    runState.eei.useGas(new BN(runState._common.param('gasPrices', 'coldsload') - baseFee))
+    // Warm (SLOAD only)
+  } else if (!isSstore) {
+    runState.eei.useGas(new BN(runState._common.param('gasPrices', 'warmstorageread') - baseFee))
+  }
+}
+
+/**
+ * Adjusts cost of SSTORE_RESET_GAS or SLOAD (aka sstorenoop) (EIP-2200) downward when storage
+ * location is already warm
+ * @param  {RunState} runState
+ * @param  {Buffer}   key          storage slot
+ * @param  {number}   defaultCost  SSTORE_RESET_GAS / SLOAD
+ * @param  {string}   costName     parameter name ('reset' or 'noop')
+ * @return {number}                adjusted cost
+ */
+function adjustSstoreGasEIP2929(
+  runState: RunState,
+  key: Buffer,
+  defaultCost: number,
+  costName: string,
+): number {
+  if (!runState._common.eips().includes(2929)) return defaultCost
+
+  const keyStr = key.toString('hex')
+  const address = runState.eei.getAddress().toString('hex')
+
+  // @ts-ignore Set Object is possibly 'undefined'
+  if (runState.accessedStorage.has(address) && runState.accessedStorage.get(address).has(keyStr)) {
+    if (costName === 'reset') {
+      return defaultCost - runState._common.param('gasPrices', 'coldsload')
+    }
+
+    if (costName === 'noop') {
+      return runState._common.param('gasPrices', 'warmstorageread')
+    }
+  }
+
+  return defaultCost
+}
+
```

Comment:
Not a concrete suggestion yet, was just wondering since the list of additional helper functions is getting so long:

Should we restructure opcode-related functionality a bit, a first suggestion would be:

- `opcodes.ts` -> `opcodes/(op)codes.ts`
- `opFns.ts` (opcode implementations) -> `opcodes/functions.ts`
- `opFns.ts` (all helper functions) -> `opcodes/util.ts`

Eventually the latter also split up:

- `opFns.ts` (EIP2929 helper functions) -> `opcodes/EIP2929.ts`

What do you guys think 🤔 ?

Reply:
@holgerd77 Agree, the helpers alone are starting to be a few hundred lines...

Happy to open that as an issue and self-assign if there's agreement... 

Would be a separate PR which we rebase this against eventually?

Reply:
I agree to split up this file 😄 , but should be a seperate PR 👍

Reply:
Yeah, makes sense to keep this separate

Reply:
This is rebased against the file re-org now...

Above it's suggested that:
> Eventually the latter also split up:
>   + opFns.ts (EIP2929 helper functions) -> opcodes/EIP2929.ts 

In 8cdc9b0 have extended this idea and moved `updateSstoreGas` out of `util`  splitting it into it's own EIP files.

Idk - this might be more than was necessary - idea is just to make `util` generic and order dynamic gas pricing logic consistently.